### PR TITLE
basic collect by line number implementation based on ast parsing

### LINKED ISSUE #3224: Allow test selection by file:lineno
`pytest tests/test_foo.py:42` should run the (nearest) test at line 42 in `tests/test_foo.py`.

When it also ignores a trailing colon it would allow for easy copy'n'pasting from failures.

Via #5973:

> To get a better overview of how it can work take a look at the [Rspec documentation](https://relishapp.com/rspec/rspec-core/v/2-3/docs/command-line/line-number-appended-to-file-path)

### PR DESCRIPTION
**Pull request by mgaitan**
Closes #3224 . 

This is a very raw implementation that find the test to collect given a line number, by traversing the code via `ast` parsing. 

Doc and tests to be added.

### CODE REVIEW

**Comment by nicoddemus** [comment]
I like the idea, thanks a lot @mgaitan! The parser implementation also is isolated and seems simple to maintain.

@bluetech @asottile @RonnyPfannschmidt @The-Compiler can we give @mgaitan the green light to continue?

File: src/_pytest/parser.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
+"""
+Find a test class and/or test method in given source code line.
+
+Original code by Pavel Savchenko
+"""
+import ast
+
+class PytestParser(ast.NodeVisitor):
```
Comment (nicoddemus) [inline src/_pytest/parser.py:8]: nitpick, but `PytestParser` doesn't tell much about what the class does, perhaps `TestLocator`?

I know you mentioned that we need to add docs already, just to highlight that we should have unittests for this class as well, in addition to 1 or more integration tests of the feature. 👍

**Comment by asottile** [comment]
at first glance this seems like a lot of complexity to take on, and it is something that will get much more complicated with time (for instance, it's incorrect for decorators and blank lines between functions right now)

there's also some things it attempts to discovery based on static information that may be indeterminable (base class checking for instance isn't really achievable statically without solving a halting problem -- the normal case of importing from a conventional place sure, but that variable could come from anywhere) -- although it looks like the `ignore_bases` code is dead code?

ast parsing is also relatively expensive (and would happen twice in this case, delaying start by a bit) -- this is probably not a _huge_ deal

there's also a lot of tricky edge cases with classes that are difficult to handle:

```python
class C:
    if True:
        def f(self): ...  # this is a method

if True:
    def f(self): ...  # this is not
```

I think the biggest problem with this is when a test is defined outside of the module being run (for example, base class tests) -- this discovery method wouldn't work at all

I guess overall I'm -.5 on this

**Comment by The-Compiler** [comment]
I'm also not sure about this. On one hand, it seems like a great path down a rabbit hole we'd maybe rather not go down. On the other hand, it doesn't need to be perfect and handle all imaginable corner cases to be useful.

**Comment by nicoddemus** [comment]
> it doesn't need to be perfect and handle all imaginable corner cases to be useful.

Indeed, but I really didn't study the patch to the length @asottile did, and given he's -5 on it I guess we should drop this then.

Thanks a lot @mgaitan for the contribution, but we will be turning this down for now. 👍

**Comment by The-Compiler** [comment]
@nicoddemus -.5 means -0.5 rather than -5 FWIW :wink:

**Comment by nicoddemus** [comment]
Ahhh I've read `-5` sorry hehehe

Well let's see if others manifest themselves, we can always reopen if others are favorable to the idea.

**Comment by mgaitan** [comment]
Let me try to defend a bit the approach, giving some answers to @asottile's points 

> at first glance this seems like a lot of complexity to take on, 

It's true that current implementation looks a bit more complicated than necessary and it needs to be polished (I've just [copied](https://github.com/asfaltboy/SublimeTestPlier/blob/develop/test_parser.py) it) , but a NodeVisitor's subclassing it's a standard, documented way to traverse the code for goals like this one. I mean, it's not an ad-hoc algorithm powered by tricky regex hard to maintain. 

> there's also some things it attempts to discovery based on static information that may be indeterminable

This is also true, but I think in this feature as a shortcut to call specific tests implemented as a function or method in an explicit way. 

For instance, if I pass a line that belong to a function that is not a test (or is one that won't be collected) that should be ok (at least in its first implementation), because the responsibility of this should be just the "mapping" and not to understand the full semantic of what the user is trying to execute. 

> ast parsing is also relatively expensive

as the main use case of this is run specific tests I guess the penalty could be dismissable. Moreover, there is a very simple improve to do, because we could ignore earlier all the nodes that ends before the given line number. 

> I think the biggest problem with this is when a test is defined outside of the module being run

I don't get this, sorry. How you could run, giving a module and a line number of that module, a test defined in other module from which the target class inherits from?  I mean, if I have  `test_base.TestBase.test_foo` defined in a module, and `test_specific.TestChild.test_bar` (where `TestChild` inherits from `TestBase`) I would be able to run  `test_specific:XX` to run `test_bar` but there is no way to run `test_foo` because there is no line number in that module that maps to that method.

**Comment by bluetech** [comment]
I think a better implementation strategy is not to go the ast route to go from line -> class/function, but to go from node -> line range and filter on that. But I haven't looked at this in detail.

Aside from that, I'm somewhat interested on what is the use case for this feature. Maybe a "Run test at this line" editor action?

**Pull request closed without merging**